vlog4j-core/src/main/java/org/semanticweb/vlog4j/core/reasoner/implementation/VLogReasoner.java

```diff
@@ -808,4 +813,59 @@ ReasonerState getReasonerState() {
 	void setReasonerState(ReasonerState reasonerState) {
 		this.reasonerState = reasonerState;
 	}
+
+	@Override
+	public Correctness writeInferences(OutputStream stream) throws IOException {
+		QueryResult queryAnswer;
+		Set<Predicate> toBeQueriedHeadPredicates = new HashSet<Predicate>();
+		TermQueryResultIterator stringQueryResultIterator;
+		QueryResultIterator answers;
+		for (Rule rule : this.knowledgeBase.getRules()) {
+			for (Literal literal : rule.getHead()) {
+				toBeQueriedHeadPredicates.add(literal.getPredicate());
+			}
+		}
+		for (DataSourceDeclaration dataSourceDeclaration : this.knowledgeBase.getDataSourceDeclarations()) {
+			toBeQueriedHeadPredicates.add(dataSourceDeclaration.getPredicate());
+		}
+		for (Fact fact : this.knowledgeBase.getFacts()) {
+			toBeQueriedHeadPredicates.add(fact.getPredicate());
+		}
+
+		for (Predicate predicate : toBeQueriedHeadPredicates) {
+			List<Term> toBeGroundedVariables = new ArrayList<Term>();
+			for (int i = 0; i < predicate.getArity(); i++) {
+				toBeGroundedVariables.add(Expressions.makeUniversalVariable("X" + i));
+			}
+			final karmaresearch.vlog.Atom vLogAtom = ModelToVLogConverter
+					.toVLogAtom(Expressions.makePositiveLiteral(predicate, toBeGroundedVariables));
+			try {
+				stringQueryResultIterator = this.vLog.query(vLogAtom, true, false);
+				answers = new VLogQueryResultIterator(stringQueryResultIterator, this.correctness);

```

Comment:
it is not necessary to do this. Why is a new VLogQueryResultIterator created? one can use the initial vlog iterator, and call `VLogToModelConverter#toTermList` on the term list that `next()` returns.